Alternate Steam Libraries

scfinniss opened the pull request:
I noticed while working on the Watcom C stuff that Chocolate Doom was not locating my IWADs.

What I've done here is scrape the alternative library locations from Steam's `libraryfolders.vdf` file and use them to add more search paths. The file itself is mostly key value pairs with a header. Here is the file from my setup for reference.

```
"LibraryFolders"
{
	"TimeNextStatsReport"		"1488133041"
	"ContentStatsID"		"198096677914456296"
	"1"		"D:\\Steam"
	"2"		"E:\\Steam"
	"3"		"C:\\Steam"
}
```

I specifically ignore `LibraryFolders`, thereafter I assume every second string is a path, as long as the preceding key is numeric. I make no real effort to understand the structure of the file beyond this.

It's important to note that the unlisted `C:\Program Files (x86)\Steam` is also a library. Presumably Steam always considers its install path a library.

I believe this may resolve issue #482 (excluding Mac and Linux support).

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
+
+static void CheckSteamLibrary(long library_file_size, char *library_file_buffer)
+{
+    long i, start;
```
Please don't use type `long`, since it means very different things in different compilers. Please remove all uses of `long`.

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
+{
+    long i, start;
+    boolean key, key_valid, value;
+    char* buffer;
```
`char *buffer;`

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
+        else if (start > -1 && key)
+        {
+            //Keys for libraries are always numeric
+            if (library_file_buffer[i] < '0' || library_file_buffer[i] > '9')
```
Please use `isdigit()`.

fragglet commented on src/d_iwad.c:
```diff
@@ -358,15 +447,53 @@ static void CheckSteamEdition(void)
+
+    if (probe != NULL)
+    {
+        library_file = fopen(probe, "rb");
```
Please factor this whole block out into a separate function.

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,97 @@ static void CheckInstallRootPaths(void)
+
+    for (i=0;i<library_file_size; ++i)
+    {
+        if (library_file_buffer[i] == '"')
```
The logic in this function is very complicated and hard to follow. It looks like you're trying to do tokenization and higher-level logic in a single function. I think you can probably simplify this greatly by splitting out these into separate concerns.

scfinniss commented:
The main change here is restructuring based on the feedback.
Changes in `CheckSteamEdition` have been mostly removed, replaced with one call to `CheckSteamLibraries`.

Then `CheckSteamLibrary` was replaced with
* `GetSteamNextLibPath`
* `IsValidSteamLibKey`
* `GetNextSteamLibToken`
This hopefully makes the logic easier to follow.

Instances of `long` replaced with `size_t`, hopefully I've not introduced any style issues this time.

scfinniss commented:
I'm going to use the [reference from Valve](https://developer.valvesoftware.com/wiki/KeyValues#File_Format) to make some changes to how I parse the `libraryfolders.vdf`.

Is it worth moving this logic to a new file? (eg `m_steam.h`, `m_steam.c`)

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
+
+    if (library_file_size != count)
+    {
+        return NULL;
```
free(library_file_buffer);

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
+        if (buffer[i] == '"')
+        {
+            //first " is the start of the token
+            if (token == NULL)
```
It becomes simpler if you just do two loops: one to find the starting ", another to find the ending ".

scfinniss replied:
Thanks, this has helped.

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
+static char *steam_libraries[MAX_STEAM_LIBRARIES];
+static int num_steam_libraries = 0;
+
+static void AddSteamLibraryDir(char *dir)
```
Do you need the `steam_libraries[]` array? Can't you just make this a wrapper around AddIWADDir?

scfinniss replied:
Yes, that array is note needed.

fragglet commented on src/d_iwad.c:
```diff
@@ -345,11 +348,198 @@ static void CheckInstallRootPaths(void)
+            return NULL;
+        }
+
+        //file always starts with this line before key/value pairs
```
So in that case, just put a call before the start of the loop in `CheckSteamLibraries()` that calls `GetNextSteamLibToken()` and checks it's equal to "LibraryFolders". Then this function should become much simpler.

scfinniss replied:
I've changed how I parse the file. I treat the whole VDF file as key value pairs, and I don't bother checking that it starts with "LibraryFolders". The code is a lot simpler now.

scfinniss commented:
I've addressed the reviews, and also made the tokenizer understand VDF more completely.

The tokenizer now handles tokens that are not always quoted, as well as ones with escaped characters.

To make the parsing simpler I've avoided the issue of subkeys completely. I ignore closing braces and treat the opening ones as part of a regular unquoted token.

So
```
"LibraryFolders"
{
	"TimeNextStatsReport"		"1505818818"
	"ContentStatsID"		"198096677914456296"
	"1"		"D:\\Steam"
	"2"		"E:\\Steam"
	"3"		"C:\\Steam"
}
```
Get tokenized to
```
"LibraryFolders"
{
"TimeNextStatsReport"
"1505818818"
"ContentStatsID"
"198096677914456296"
"1"
"D:\\Steam"
"2"
"E:\\Steam"
"3"
"C:\\Steam"
```
Every odd token is a key, every even token is a value.

fragglet commented on src/d_iwad.c:
```diff
@@ -345,27 +348,245 @@ static void CheckInstallRootPaths(void)
+        break;
+    }
+
+    for (; i<buff_size; ++i)
```
This ends up being complicated because you're trying to process different types of things using the same loop. You'd be better off just treating them separately. For example:

```
c = ... whatever
if (c == '"') {
    return ReadStringToken(.....);
} else if (c == '{') {
    return TOKEN_OPEN_BRACE;
} else if (c == '}') {
    return TOKEN_CLOSE_BRACE;
} else if (isalpha(c)) {
    return ReadKeyword(......);
} else {
    return TOKEN_UNKNOWN;
}

scfinniss commented on src/d_iwad.c:
```diff
@@ -345,27 +348,245 @@ static void CheckInstallRootPaths(void)
+//Read whole of libraryfolders.vdf as string
+//silently fail is something goes wrong reading the file
+
+static char *GetSteamLibText(char *library_path)
```
Yes, I've got it working with `M_ReadFile()`.

The first difference I see between `M_ReadFile()` and my code is that I'm adding the `\0` to the end of the buffer. I assume I did that to ensure `strlen()` is guaranteed to work later. Not sure if that was ever necessary.

The other difference is that I'm using `malloc()`, where `M_ReadFile()` uses `Z_Malloc()`.

There are a few places where `malloc()` is used over `Z_Malloc()`, are these intentional or should `Z_Malloc()` always be used?

mfrancis95 commented:
@scfinniss any updates?

Otherwise I don't think it's good to keep a nearly 3 year-old PR up.

scfinniss commented:
Real life got between me and finishing this, and I guess I forgot about it after awhile.

I'm not sure how much Chocolate Doom has changed since I worked on this, but I don't mind looking into it again if there's still interest in this feature?

mfrancis95 commented:
There have always been efforts to improve WAD-locating, even in recent times. I would say it's still worth it, but it just happens to be a fairly sizeable undertaking.
